File: app/src/main/java/io/github/hiro/lime/hooks/RemoveAds.java
package io.github.hiro.lime.hooks;

import android.graphics.Canvas;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;

import androidx.annotation.NonNull;

import java.lang.reflect.Method;

import io.github.hiro.lime.Constants;
import io.github.hiro.lime.LimeModule;
import io.github.hiro.lime.LimeOptions;
import io.github.libxposed.api.XposedInterface;

public class RemoveAds implements IHook {

    @Override
    public void hook(LimeModule module, ClassLoader classLoader, LimeOptions limeOptions) throws Throwable {
        
        // 1. 攔截廣告請求
        try {
            Class<?> requestClass = classLoader.loadClass(Constants.REQUEST_HOOK.className);
            for (Method method : requestClass.getDeclaredMethods()) {
                if (method.getName().equals(Constants.REQUEST_HOOK.methodName)) {
                    module.hook(method, new XposedInterface.Hooker() {
                        @Override
                        public Object intercept(@NonNull XposedInterface.BeforeHookCallback callback) throws Throwable {
                            Object[] args = callback.getArgs();
                            if (args != null && args.length > 0 && args[0] != null) {
                                String request = args[0].toString();
                                if (request.equals("getBanners") || request.equals("getPrefetchableBanners")) {
                                    // 強制回傳 null，阻擋廣告請求
                                    callback.setResult(null);
                                    return null; 
                                }
                            }
                            return callback.callOriginal();
                        }
                    });
                }
            }
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "RemoveAds (Request) 失敗: " + e.getMessage());
        }

        // 2. 隱藏 SmartChannel (頂部廣告)
        try {
            Class<?> smartChannelClass = classLoader.loadClass("com.linecorp.line.admolin.smartch.v2.view.SmartChannelViewLayout");
            Method dispatchDrawMethod = smartChannelClass.getDeclaredMethod("dispatchDraw", Canvas.class);
            
            module.hook(dispatchDrawMethod, new XposedInterface.Hooker() {
                @Override
                public Object intercept(@NonNull XposedInterface.BeforeHookCallback callback) throws Throwable {
                    View view = (View) callback.getThisObject();
                    if (view != null && view.getParent() instanceof View) {
                        ((View) view.getParent()).setVisibility(View.GONE);
                    }
                    return callback.callOriginal();
                }
            });
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "RemoveAds (SmartChannel) 失敗: " + e.getMessage());
        }

        // 3. 隱藏 LadAdView (貼文串廣告)
        try {
            Class<?> ladAdViewClass = classLoader.loadClass("com.linecorp.line.ladsdk.ui.common.view.lifecycle.LadAdView");
            Method onAttachedMethod = ladAdViewClass.getDeclaredMethod("onAttachedToWindow");
            
            module.hook(onAttachedMethod, new XposedInterface.Hooker() {
                @Override
                public Object intercept(@NonNull XposedInterface.BeforeHookCallback callback) throws Throwable {
                    View view = (View) callback.getThisObject();
                    if (view != null && view.getParent() != null && view.getParent().getParent() instanceof View) {
                        View grandParent = (View) view.getParent().getParent();
                        ViewGroup.LayoutParams layoutParams = grandParent.getLayoutParams();
                        if (layoutParams != null) {
                            layoutParams.height = 0;
                            grandParent.setLayoutParams(layoutParams);
                        }
                        grandParent.setVisibility(View.GONE);
                    }
                    return callback.callOriginal();
                }
            });
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "RemoveAds (LadAdView) 失敗: " + e.getMessage());
        }

        // 4. 動態攔截帶有 "Ad" 的 View
        try {
            Method addViewMethod = ViewGroup.class.getDeclaredMethod("addView", View.class, ViewGroup.LayoutParams.class);
            module.hook(addViewMethod, new XposedInterface.Hooker() {
                @Override
                public Object intercept(@NonNull XposedInterface.BeforeHookCallback callback) throws Throwable {
                    Object result = callback.callOriginal(); // 先讓它執行完
                    Object[] args = callback.getArgs();
                    if (args != null && args.length > 0 && args[0] instanceof View) {
                        View view = (View) args[0];
                        String className = view.getClass().getName();
                        if (className.contains("Ad")) {
                            view.setVisibility(View.GONE);
                        }
                    }
                    return result;
                }
            });
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "RemoveAds (addView) 失敗: " + e.getMessage());
        }

        // 5. WebView JS 注入
        try {
            Class<?> webViewClientClass = classLoader.loadClass(Constants.WEBVIEW_CLIENT_HOOK.className);
            for (Method method : webViewClientClass.getDeclaredMethods()) {
                if (method.getName().equals(Constants.WEBVIEW_CLIENT_HOOK.methodName)) {
                    Class<?>[] paramTypes = method.getParameterTypes();
                    if (paramTypes.length == 2 && paramTypes[0] == WebView.class && paramTypes[1] == String.class) {
                        module.hook(method, new XposedInterface.Hooker() {
                            @Override
                            public Object intercept(@NonNull XposedInterface.BeforeHookCallback callback) throws Throwable {
                                Object result = callback.callOriginal();
                                Object[] args = callback.getArgs();
                                if (args != null && args.length > 0 && args[0] instanceof WebView) {
                                    WebView webView = (WebView) args[0];
                                    webView.evaluateJavascript("(() => { /* JS 內容 */ })();", null);
                                }
                                return result;
                            }
                        });
                    }
                }
            }
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "RemoveAds (WebView) 失敗: " + e.getMessage());
        }
    }
}

File: app/src/main/java/io/github/hiro/lime/hooks/RemoveAds.kt
package io.github.hiro.lime.hooks

import android.graphics.Canvas
import android.util.Log
import android.view.View
import android.view.ViewGroup
import io.github.hiro.lime.LimeModule
import io.github.hiro.lime.LimeOptions
import io.github.libxposed.api.XposedModuleInterface.ExceptionMode

class RemoveAds : IHook {

    override fun hook(module: LimeModule, classLoader: ClassLoader, options: LimeOptions) {
        
        // 1. 攔截廣告請求 (Thrift Request 層)
        runCatching {
            val requestClass = classLoader.loadClass(Constants.REQUEST_HOOK.className)
            val reqMethod = requestClass.getDeclaredMethod(
                Constants.REQUEST_HOOK.methodName, 
                Any::class.java, 
                Any::class.java
            )

            module.hook(reqMethod, module.newHookBuilder()
                .setInterceptor { chain ->
                    val args = chain.args
                    if (args.isNotEmpty() && args[0]?.toString()?.contains("Banners") == true) {
                        module.log(Log.DEBUG, "LIMEs", "🚫 已阻斷廣告請求: ${args[0]}")
                        return@setInterceptor null // 直接擋下廣告 Request，不執行 proceed
                    }
                    chain.proceed()
                }.build()
            )
        }.onFailure { e -> module.log(Log.ERROR, "LIMEs", "RemoveAds (Request) 失敗: ${e.message}") }

        // 2. 隱藏頂部廣告 SmartChannel (View 繪製層)
        runCatching {
            val smartChannelClass = classLoader.loadClass("com.linecorp.line.admolin.smartch.v2.view.SmartChannelViewLayout")
            val dispatchDrawMethod = smartChannelClass.getDeclaredMethod("dispatchDraw", Canvas::class.java)

            module.hook(dispatchDrawMethod, module.newHookBuilder()
                .setInterceptor { chain ->
                    val view = chain.thisObject as? View
                    // 找到 SmartChannelView 的父容器並將其設為 GONE
                    (view?.parent as? View)?.visibility = View.GONE
                    chain.proceed()
                }.build()
            )
        }.onFailure { e -> module.log(Log.ERROR, "LIMEs", "RemoveAds (SmartChannel) 失敗: ${e.message}") }

        // 3. 隱藏動態產生的 Ad View (Layout 佈局層)
        runCatching {
            val addViewMethod = ViewGroup::class.java.getDeclaredMethod(
                "addView", 
                View::class.java, 
                ViewGroup.LayoutParams::class.java
            )

            module.hook(addViewMethod, module.newHookBuilder()
                .setPriority(Int.MIN_VALUE) // 較低優先權，確保在 View 被加入後處理
                .setInterceptor { chain ->
                    val result = chain.proceed()
                    val args = chain.args
                    
                    if (args.isNotEmpty()) {
                        val v = args[0] as? View
                        // 檢查 View 的類別名稱是否包含 "Ad"，如果是則隱藏
                        if (v?.javaClass?.name?.contains("Ad") == true) {
                            v.visibility = View.GONE
                            // module.log(Log.DEBUG, "LIMEs", "🕵️ 已隱藏動態廣告元件: ${v.javaClass.name}")
                        }
                    }
                    result
                }.build()
            )
        }.onFailure { e -> module.log(Log.ERROR, "LIMEs", "RemoveAds (addView) 失敗: ${e.message}") }
    }
}
